libcaf_net/caf/net/http/client_factory.hpp

```diff
@@ -0,0 +1,174 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/net/checked_socket.hpp"
+#include "caf/net/dsl/client_factory_base.hpp"
+#include "caf/net/http/async_client.hpp"
+#include "caf/net/http/client.hpp"
+#include "caf/net/http/config.hpp"
+#include "caf/net/http/response.hpp"
+#include "caf/net/octet_stream/transport.hpp"
+#include "caf/net/ssl/connection.hpp"
+#include "caf/net/ssl/transport.hpp"
+#include "caf/net/tcp_stream_socket.hpp"
+
+#include "caf/async/future.hpp"
+#include "caf/async/promise.hpp"
+#include "caf/async/spsc_buffer.hpp"
+#include "caf/detail/latch.hpp"
+#include "caf/disposable.hpp"
+#include "caf/timespan.hpp"
+
+#include <chrono>
+#include <cstdint>
+#include <functional>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+namespace caf::net::http {
+
+/// Factory for the `with(...).connect(...).start(...)` DSL.
+class CAF_NET_EXPORT client_factory
+  : public dsl::client_factory_base<client_config, client_factory> {
+public:
+  using super = dsl::client_factory_base<client_config, client_factory>;
+
+  using config_type = client_config;
```

Comment:
The `client_config` is built with an eager setup in mind, where the user can provide a socket, ready-to-use connection or only the credentials. I simply wouldn't use it here.

We'll have either a lazy config (URI + optional context factory) or an error. Here's an outline for what I mean:

```c++
struct client_config {
  struct lazy { // Note: *not* deriving `has_ctx` here.
    /// The locator for the remote resource (HTTP or HTTPS).
    uri locator;

    /// Factory function for creating an SSL/TLS context when using HTTPS.
    std::function<...> ctx_factory;

    /// The delay between connection attempts.
    timespan retry_delay = std::chrono::seconds{1};

    /// The timeout when trying to connect.
    timespan connection_timeout = infinite;

    /// The maximum amount of retries.
    size_t max_retry_count = 0;
  };

  ...
};

/// inside client_factory

  /// Sets the optional SSL context factory.
  /// @param ctx_fn The function creating the SSL context for encryption.
  /// @returns a reference to `*this`.
  template <class Factory>
  client_factory& context_factory(Factory&& factory) {
    if (auto* lazy = get_if<client_config::lazy>(config().data))
      lazy->ctx_factory = std::move(factory);
    return *this;
  }
```

Probably means we need to copy the few functions from `caf::net::dsl::client_factory_base` over (`retry_delay` etc.) but I wouldn't mind that too much. I also not super happy with the entire DSL stuff myself, btw. At some point we need to make a pass over it and simplify it.

Reply:
I'm kind of against specializing this feature only to HTTP. 
- It wouldn't really play nice with the idea of the client being able to use both interfaces (`.context` and `.context_factory`). 
- If we don't inherit `has_ctx` then the client won't have the `shared_ptr<ssl::context` ctx`, and special functions that come from it like `connection_with_ctx`, so the code duplication would be considerable.
- We would have to do this setup again if we wanted to introduce `has_lazy_context` to websocket (`ws` vs `wss`).
Could we think of any other way to refactor this? I could factor out the factory stuff from `has_ctx` into `has_ctx_factory` and include it everywhere? All in all, I agree that the DSL stuff could use some refactoring for clarity.

Reply:
> It wouldn't really play nice with the idea of the client being able to use both interfaces (`.context` and `.context_factory`).

My mental model of these two is that one is eager and one is lazy. You cannot be both, which is why I'm unhappy with `has_ctx` trying to be both. These two are opposites, pick one or the other.

If do want to keep things uniformly, then we could make `has_ctx` lazy instead and not have an eager version at all. Then, `has_ctx` would only have one member: `make_ctx`. The eager `.context()` version would then simply set `make_ctx` to return the eagerly constructed object while the `.context_factory()` version would override the `make_ctx` member with whatever factory the user provides. A default-constructed `make_ctx` member would then mean "no SSL".

> If we don't inherit `has_ctx` then the client won't have the `shared_ptr<ssl::context> ctx`, and special functions that come from it like `connection_with_ctx`, so the code duplication would be considerable.

We could always just do something like `has_ctx{std::make_shared<ssl::context>(factory())}.connection_with_ctx(...)` if we need to at the point where we lazily spin up everything. However, I think there's not much code duplication, because `connection_with_ctx` really just wraps `ctx->new_connection(fd)` with a few if-else branches to check whether we so SSL or not plus error handling.

> We would have to do this setup again if we wanted to introduce has_lazy_context to websocket (`ws` vs `wss`).

Wouldn't the `client_config` we implement here be generic enough to push it to the `dsl` namespace as a followup?

I think either that or make `has_ctx` be always lazy. Both works for me.

Reply:
I was also drawing the parallel from the `base_config` used by the `with_t` - alias from `generic_config_value`, which has `has_ctx` as well and transfers the information to the client or server config. In that way it should have both versions. I agree it can't be both at the same time, but needs to support both.  
After some tryouts, what I'd find best would be to displace the `has_ctx` from lazy, server, conn, ... variant types to the config directly.

Reply:
Done. Reworked the lazy parts.